src/python/pants/rules/core/list_target_types_test.py

```diff
@@ -0,0 +1,117 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from contextlib import contextmanager
+from textwrap import dedent
+from typing import Optional, cast
+from unittest.mock import Mock
+
+from pants.engine.rules import UnionMembership
+from pants.engine.target import BoolField, RegisteredTargetTypes, StringField, Target
+from pants.rules.core.list_target_types import list_target_types
+from pants.testutil.engine.util import MockConsole, run_rule
+from pants.util.ordered_set import OrderedSet
+
+
+# TODO(#9141): replace this with a proper util to create `GoalSubsystem`s
+class MockOptions:
+    def __init__(self, **values):
+        self.values = Mock(**values)
+
+    @contextmanager
+    def line_oriented(self, console: MockConsole):
+        yield lambda msg: console.print_stdout(msg)
+
+
+# Note no docstring.
+class GhcVersion(StringField):
+    alias = "ghc_version"
+
+
+class HaskellLibrary(Target):
+    """A library of Haskell code."""
+
+    alias = "haskell_library"
+    core_fields = (GhcVersion,)
+
+
+# Note multiline docstring.
+class HaskellTests(Target):
+    """Tests for Haskell code.
+
+    This assumes that you use QuickCheck or an equivalent test runner.
+    """
+
+    alias = "haskell_tests"
+    core_fields = (GhcVersion,)
+
+
+# Note no docstring.
+class HaskellBinary(Target):
+    alias = "haskell_binary"
+    core_fields = (GhcVersion,)
+
+
+def run_goal(
+    *, union_membership: Optional[UnionMembership] = None, details_target: Optional[str] = None
+) -> str:
+    console = MockConsole(use_colors=False)
+    run_rule(
+        list_target_types,
+        rule_args=[
+            RegisteredTargetTypes.create([HaskellBinary, HaskellLibrary, HaskellTests]),
+            union_membership or UnionMembership({}),
+            MockOptions(details=details_target),
+            console,
+        ],
+    )
+    return cast(str, console.stdout.getvalue())
+
+
+def test_list_all() -> None:
+    stdout = run_goal()
+    assert (
+        stdout
+        == """\
+                haskell_binary: <no description>
+               haskell_library: A library of Haskell code.
+                 haskell_tests: Tests for Haskell code.\n"""
+    )
```

Comment:
NB: I originally used a looser collection of `assert haskell_binary in stdout` style checks. I replaced it for this strict string equality because it allows us to guarantee things like correct leading whitespace and sorted output.

These tests are 100% isolated thanks to defining these custom Haskell target types, so the tests will only need to change when we change `list_target_types.py`. I think that's an acceptable amount of fragility here.